pkg/pillar/cmd/verifier/verifier.go

```diff
@@ -1261,55 +879,229 @@ func handleGlobalConfigDelete(ctxArg interface{}, key string,
-	publications := []pubsub.Publication{
-		ctx.pubAppImgPersistStatus,
-		ctx.pubBaseOsPersistStatus,
+// Recreate status files for verified objects as types.DOWNLOADED
+// except containers which we mark as types.VERIFIED
+func handleInitVerifiedObjects(ctx *verifierContext) {
+
+	for _, objType := range verifierObjTypes {
+
+		verifiedDirname := getVerifiedDir(objType)
+		if _, err := os.Stat(verifiedDirname); err == nil {
+			populateInitialStatusFromVerified(ctx, objType,
+				verifiedDirname, "")
+		}
 	}
-	for _, pub := range publications {
-		items := pub.GetAll()
-		for _, st := range items {
-			status := st.(types.PersistImageStatus)
-			if status.RefCount == 0 {
-				status.LastUse = time.Now()
-				log.Infof("gcResetObjectLastUse: reset %v LastUse to now", status.Key())
-				publishPersistImageStatus(ctx, &status)
+}
+
+func persistImageStatusFromImageFile(
+	objType, objDirname, parentDirname, imageFileName string,
+	size int64, filename string) *types.PersistImageStatus {
+
+	status := types.PersistImageStatus{
+		VerifyStatus: types.VerifyStatus{
+			Name:         imageFileName,
+			ObjType:      objType,
+			FileLocation: filename,
+			ImageSha256:  parentDirname,
+			Size:         size,
+		},
+		LastUse:  time.Now(),
+		RefCount: 0,
+	}
+
+	log.Debugf("persistImageStatusFromImageFile: objType: %s, "+
+		"objDirname: %s, parentDirname: %s, imageFileName: %s",
+		objType, objDirname, parentDirname, imageFileName)
+
+	log.Debugf("persistImageStatusFromImageFile: status: %+v", status)
+	return &status
+}
+
+// Recursive scanning for verified objects,
+// to recreate the VerifyImageStatus.
+func populateInitialStatusFromVerified(ctx *verifierContext,
+	objType string, objDirname string, parentDirname string) {
+
+	log.Infof("populateInitialStatusFromVerified(%s, %s)", objDirname,
+		parentDirname)
+
+	locations, err := ioutil.ReadDir(objDirname)
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	log.Debugf("populateInitialStatusFromVerified: processing locations %v", locations)
+
+	for _, location := range locations {
+
+		filename := objDirname + "/" + location.Name()
+
+		if location.IsDir() {
+			log.Debugf("populateInitialStatusFromVerified: Looking in %s", filename)
+			if _, err := os.Stat(filename); err == nil {
+				populateInitialStatusFromVerified(ctx,
+					objType, filename, location.Name())
+			}
+		} else {
+			size := int64(0)
+			info, err := os.Stat(filename)
+			if err != nil {
+				// XXX Delete file?
+				log.Error(err)
+			} else {
+				size = info.Size()
+			}
+			log.Debugf("populateInitialStatusFromVerified: Processing %s: %d Mbytes",
+				filename, size/(1024*1024))
+			status := persistImageStatusFromImageFile(objType, objDirname,
+				parentDirname, location.Name(), size, filename)
+			if status != nil {
+				publishPersistImageStatus(ctx, status)
 			}
 		}
 	}
 }
+
+func getVerifierDir(objType string) string {
+	return types.DownloadDirname + "/" + objType + "/" + "verifier"
```

Comment:
Can we lock this down to inside verifier? I went through the PR, it _looks_ like the only consumer of `types.DownloadDirName` is inside verifier (which is good), but that still means it is known and exported, and we wrap it all entirely inside `getVerifierDir()` and `getVerifiedDir()`, which is great.

Can we make it not even known outside, which is the goal? 

I was thinking something like this:

* verifier has a startup option `controlledPath` (I don't care if it is env var or CLI or both, or if it has a better name). The default can be the current, e.g. `/persist/....`
* verifier creates its own `verified/` and `verifier/` directories under there

Essentially, verifier on startup is given a path where it knows it is in complete and total control; nothing else should write to there, ideally nothing else even _can_ write to there, but others can read from there. Verifier is at liberty to move files around in there, create, delete, use any file structure path it wants.

The **only** contract it has with the outside world is that once it says a file is verified, and it has published that file location, it cannot move that file. It might delete it, but that is the only place that file will exist. The outside world (mainly volumemgr) has a right to expect that file path will be the path to that verified file, unless it is deleted. Not the directory, it can make no assumptions about the directory, but that path.

Short form: this looks right, very close, we should pull it not step tighter to verifier's control.

Simple test: can this PR remove `types.DownloadDirName`? It probably cannot, as volumemgr depends on it, but in theory, at least as far as verifier is concerned, it could.

Reply:
I thought of a different test that might work as well. It might even be necessary.

Can we have the verifier and verified in a different directory?

Currently, it is all in 

```
/persist/downloads/<type>/pending/
/persist/downloads/<type>/verifier/
/persist/downloads/<type>/verified/
```

For this to work correctly, the first needs to be under volumemgr control, the latter two under verifier control, yet there is a common root to all three, making it difficult to split. 

When this PR is in, we should be able to have the dirs be:

```
/persist/downloads/<type>/pending/
/persist/downloads/verifier/special/controlled/donttouch/<type>/verifier/
/persist/downloads/verifier/special/controlled/donttouch/<type>/verified/
```

or any other path, and nothing else is broken. 

In the above, I would run verifier by telling it (CLI or env var), "your controlled root is `/persist/downloads/verifier/special/controlled/donttouch/`""

Reply:
They can even be in /persist/verifier/ if you'd like. Nothing special with /persist/downloads

Reply:
Added a constant in Verifier service which points to the path `/persist/verifier/`.
Now only Verifier service will be aware of this path.

Reply:
I like this, quite a bit.

Once this is in, I will be happy to add a simple PR to add an option to change the dir, which will make it easier to run locally.

@eriknordmark , is there any chance this will break anything already downloaded and verified? Do we need any special handling for it? I 100% agree that this should be under verifier control, and it should be arbitrary, precisely as @adarsh-zededa did.

Might we have a lingering file in `/persist/downloads/appImg.obj/verified/foo/` that we now no longer reference? 

My concerns are twofold:

* we might not know about a file and thus re-download it
* we might waste space with existing ones that no one knows to clean up

If that is not a concern, then don't worry about it. If it is, we might want to add some logic to allow volumemgr to send those over for verification. But I would do it in a separate PR (if it even is needed). Or maybe just a general garbage collection?

Reply:
@deitch We might need to check that any WIP download in pending is deleted on a device reboot to avoid confusion on that side. Ditto if files sit in the verifier directory from a device reboot/crash (I think there is code to handle the verifier side but I haven't double-checked).

We can do those checks separately from this PR.

Reply:
@nordmark for the files in pending and verifier directories, `eve/pkg/pillar/cmd/volumemgr/dirs.go: clearInProgressDownloadDirs()` and  `eve/pkg/pillar/cmd/verifier/dirs.go: clearInProgressDownloadDirs()` respectively handles cleaning up the files after device reboot.